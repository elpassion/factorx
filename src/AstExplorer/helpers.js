// @flow
import { sortBy } from 'lodash';

export const referencePathsForVariableInScope = (scope: Object, variableName: string) => {
  const binding = scope.getOwnBinding(variableName);
  let referencePaths = [];
  scope.path.traverse({
    AssignmentExpression: (assignmentPath) => {
      if (assignmentPath.node.left.name === variableName) {
        assignmentPath.traverse({
          Identifier: (identifierPath) => {
            if (identifierPath.key === 'left' && identifierPath.parentPath === assignmentPath) {
              referencePaths.push(identifierPath);
            }
          },
        });
      }
    },
  });
  let bindingPaths = binding.referencePaths;

  bindingPaths = bindingPaths.filter(refPath => refPath.node !== binding.identifier);
  referencePaths = referencePaths.concat(bindingPaths);
  return sortBy(referencePaths, [path => path.node.start]);
};

export const findScopeRoad = (scope: Object) => {
  const scopeRoad = [];
  scope.path.find((path) => {
    scopeRoad.push(path.key);
    return path.isProgram();
  });
  return scopeRoad;
};
