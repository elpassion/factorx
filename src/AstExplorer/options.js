// @flow
import babylon from 'recast';

const options = {
  parser: {
    parse(source: string) {
      return babylon.parse(source, {
        plugins: [
          'asyncGenerators',
          'classConstructorCall',
          'classProperties',
          'decorators',
          'doExpressions',
          'exportExtensions',
          'flow',
          'functionSent',
          'functionBind',
          'jsx',
          'objectRestSpread',
          'dynamicImport',
        ],
      });
    },
  },
};

export default options;
