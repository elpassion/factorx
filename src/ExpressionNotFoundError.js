// @flow

export default class ExpressionNotFoundError extends Error {
  constructor(message?: string) {
    super(message);
    this.name = 'ExpressionNotFoundError';
  }
}
