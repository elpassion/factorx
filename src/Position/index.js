// @flow

export default class Position {
  start: number;
  end: number;

  constructor(start: number, end: number) {
    this.start = start;
    this.end = end;
  }

  static fromNode(node: Object) {
    return new Position(node.start, node.end);
  }

  isBlank() {
    return this.start === this.end;
  }

  equalToPosition(otherPosition: Position) {
    return otherPosition.start === this.start && otherPosition.end === this.end;
  }

  includesPosition(otherPosition: Position) {
    return otherPosition.start <= this.start && otherPosition.end >= this.end;
  }

  includes(otherPosition: Position) {
    return (this.isBlank() && this.includesPosition(otherPosition)) ||
      this.equalToPosition(otherPosition);
  }
}
