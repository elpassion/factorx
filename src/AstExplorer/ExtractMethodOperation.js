// @flow
import recast from 'recast';
import traverse from 'babel-traverse';
import * as types from 'babel-types';
import findIndex from 'lodash/findIndex';
import Position from '../Position';
import ExpressionNotFoundError from '../ExpressionNotFoundError';
import options from './options';

export default class ExtractMethodOperation {
  ast: Object;
  code: string;
  replacedNodesCount: number;
  scopeRoad: Array<any>;
  variableIdentifier: Object;
  selections: Array<Position>;
  selection: Position;
  cursorPositions: Array<Position>;

  constructor(ast: Object, selections: Array<Position>) {
    this.ast = ast;
    this.replacedNodesCount = 0;
    this.scopeRoad = [];
    this.selections = selections;
    this.selection = selections[0];
    this.cursorPositions = [];
    this.variableIdentifier = types.identifier('extracted');
  }

  getPathScope = (path: Object) =>
    (types.isArrowFunctionExpression(path) ? path.scope.parent : path.scope);

  findDeepestIncludingPath = (programPath: Object) => {
    const state = {};
    programPath.traverse({
      enter: (path) => {
        if (this.selection.includesPosition(Position.fromNode(path.node))) {
          state.deepestIncludedPath = path;
        } else {
          path.skip();
        }
      },
    });
    return state.deepestIncludedPath || programPath;
  };

  cloneSelectedNodes = (pathNodes: Array<Object>) => {
    let nodes = [...pathNodes];
    const startNodeIndex = findIndex(nodes, {
      start: this.selection.start,
    });
    const endNodeIndex =
      findIndex(nodes, {
        end: this.selection.end,
      }) + 1;
    nodes = nodes.slice(startNodeIndex, endNodeIndex);
    return nodes.map(node => types.clone(node));
  };

  buildExpressionStatement = () => {
    const callExpression = types.callExpression(this.variableIdentifier, []);
    return types.expressionStatement(callExpression);
  };

  removeSelectedNodes = (deepestIncludedPath: Object, clonedNodes: Array<Object>) => {
    let lastClonedNodeIndex = 0;
    const isExactlyTheFirstNode = otherNode =>
      types.isNodesEquivalent(clonedNodes[lastClonedNodeIndex], otherNode) &&
      Position.fromNode(clonedNodes[lastClonedNodeIndex]).equalToPosition(
        Position.fromNode(otherNode),
      );
    const statement = this.buildExpressionStatement();
    let insertedInvocation = false;
    if (!deepestIncludedPath) throw new ExpressionNotFoundError();
    deepestIncludedPath.traverse({
      enter: (path) => {
        if (isExactlyTheFirstNode(path.node)) {
          if (!insertedInvocation) {
            path.insertBefore(statement);
            insertedInvocation = true;
          }
          path.remove();
          lastClonedNodeIndex += 1;
        }
      },
    });
  };

  buildVariableDeclaration = (body) => {
    const functionExpression = types.arrowFunctionExpression([], body);
    return { id: this.variableIdentifier, init: functionExpression, kind: 'const' };
  };

  extractMultipleNodes = (deepestIncludedPath: Object) => {
    const scope = this.getPathScope(deepestIncludedPath);
    const nodes = this.cloneSelectedNodes(deepestIncludedPath.node.body);
    this.removeSelectedNodes(deepestIncludedPath, nodes);
    scope.push(this.buildVariableDeclaration(types.blockStatement(nodes)));
  };

  extractExpression = (deepestIncludedPath: Object) => {
    const scope = this.getPathScope(deepestIncludedPath);
    const node = deepestIncludedPath.node;
    scope.push(this.buildVariableDeclaration(types.clone(node)));
    deepestIncludedPath.replaceWith(this.buildExpressionStatement());
  };

  extractStatement = (deepestIncludedPath: Object) => {
    const scope = this.getPathScope(deepestIncludedPath);
    const node = types.blockStatement([types.clone(deepestIncludedPath.node)]);
    scope.push(this.buildVariableDeclaration(types.clone(node)));
    deepestIncludedPath.replaceWith(this.buildExpressionStatement());
  };

  extractMethod = (programPath: Object) => {
    const deepestIncludingPath = this.findDeepestIncludingPath(programPath);
    if (types.isBlock(deepestIncludingPath)) {
      this.extractMultipleNodes(deepestIncludingPath);
    } else if (types.isExpression(deepestIncludingPath)) {
      this.extractExpression(deepestIncludingPath);
    } else if (types.isStatement(deepestIncludingPath)) {
      this.extractStatement(deepestIncludingPath);
    }
  };

  start(): { ast: Object, result: { code: string, cursorPositions: Array<Position> } } {
    this.transform(this.extractMethod);

    return { ast: this.ast, result: { code: this.code, cursorPositions: this.cursorPositions } };
  }

  transform(transformation: Function) {
    traverse(this.ast, {
      Program(programPath) {
        transformation(programPath);
      },
    });
    this.code = recast.print(this.ast).code;
    this.ast = recast.parse(this.code, options);
  }
}
