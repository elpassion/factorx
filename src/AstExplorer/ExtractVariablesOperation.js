// @flow
import recast from 'recast';
import traverse from 'babel-traverse';
import * as types from 'babel-types';
import isEqual from 'lodash/isEqual';
import Position from '../Position';
import ExpressionNotFoundError from '../ExpressionNotFoundError';
import options from './options';

export default class ExtractVariablesOperation {
  ast: Object;
  code: string;
  replacedNodesCount: number;
  scopeRoad: Array<any>;
  variableType: 'const' | 'let';
  variableIdentifier: Object;
  selections: Array<Position>;
  cursorPosition: Position;

  constructor(ast: Object, variableType: 'const' | 'let' = 'let', selections: Array<Position>) {
    this.ast = ast;
    this.replacedNodesCount = 0;
    this.scopeRoad = [];
    this.variableType = variableType;
    this.selections = selections;
  }

  isAllSelectionsExtracted() {
    return this.replacedNodesCount !== this.selections.length;
  }

  isSelectedExpression(expressionPath: Object) {
    const isExpressionIncluded = selection =>
      selection.includes(Position.fromNode(expressionPath.node));

    const isExpressionTheDeclaration = expressionPath.parent &&
      expressionPath.parent.type === 'VariableDeclarator' &&
      expressionPath.parent.id === this.variableIdentifier;

    return this.selections.find(
      selection => isExpressionIncluded(selection) && !isExpressionTheDeclaration,
    );
  }

  replaceSelectionsInScope(scope: Object) {
    scope.path.traverse({
      Expression: (expressionPath) => {
        if (this.isSelectedExpression(expressionPath)) {
          expressionPath.replaceWith(types.identifier(this.variableIdentifier.name));
          this.replacedNodesCount += 1;
        }
      },
    });
  }

  addDeclarationToScope(scope: Object, node: Object) {
    this.variableIdentifier = scope.generateUidIdentifierBasedOnNode(node.id);
    scope.push({ id: this.variableIdentifier, init: node, kind: this.variableType });
  }

  // eslint-disable-next-line class-methods-use-this
  findScopeRoad(scope: Object) {
    const scopeRoad = [];
    scope.path.find((path) => {
      scopeRoad.push(path.key);
      return types.isProgram(path);
    });
    return scopeRoad;
  }

  saveScopeRoad(scope: Object) {
    this.scopeRoad = this.findScopeRoad(scope);
  }

  extractVariable = (programPath: Object) => {
    const firstSelection = this.selections[0];
    programPath.traverse({
      Expression: (path) => {
        const { node, scope } = path;
        if (!node.visited && firstSelection.includes(Position.fromNode(node))) {
          node.visited = true;
          this.addDeclarationToScope(scope, node);
          this.replaceSelectionsInScope(scope);
          this.saveScopeRoad(scope);
          if (this.isAllSelectionsExtracted()) programPath.stop();
        }
      },
    });

    if (this.isAllSelectionsExtracted()) throw new ExpressionNotFoundError();
  };

  moveDeclaration = (programPath: Object) => {
    const traversalState = {};
    programPath.traverse(
      {
        VariableDeclarator: (declaratorPath, state) => {
          if (
            isEqual(this.findScopeRoad(declaratorPath.scope), this.scopeRoad) &&
            declaratorPath.node.id.name === this.variableIdentifier.name
          ) {
            const { scope } = declaratorPath;
            const binding = scope.bindings[this.variableIdentifier.name];
            const declarationPath = declaratorPath.parentPath;
            const firstReferencePath = binding.referencePaths[0];
            const firstReferencePathParentStatement = firstReferencePath.find(
              path =>
                types.isStatement(path) &&
                types.isNodesEquivalent(path.scope.path.node, scope.path.node),
            );
            //eslint-disable-next-line
            state.firstReferencePathParentStatement = firstReferencePathParentStatement;
            //eslint-disable-next-line
            state.replacement = types.clone(declarationPath.node);
            declarationPath.remove();
            programPath.stop();
          }
        },
      },
      traversalState,
    );
    traversalState.firstReferencePathParentStatement.insertBefore(traversalState.replacement);
  };

  setCursorPosition = (programPath: Object) => {
    programPath.traverse({
      VariableDeclarator: (declaratorPath) => {
        if (
          isEqual(this.findScopeRoad(declaratorPath.scope), this.scopeRoad) &&
          declaratorPath.node.id.name === this.variableIdentifier.name
        ) {
          const { scope } = declaratorPath;
          this.cursorPosition = Position.fromNode(
            scope.bindings[this.variableIdentifier.name].path.node.id,
          );
          programPath.stop();
        }
      },
    });
  };

  start(): { ast: Object, result: { code: string, cursorPosition: Position | typeof undefined } } {
    this.transform(this.extractVariable);
    this.transform(this.moveDeclaration);
    this.transform(this.setCursorPosition);

    return { ast: this.ast, result: { code: this.code, cursorPosition: this.cursorPosition } };
  }

  transform(transformation: Function) {
    traverse(this.ast, {
      Program(programPath) {
        transformation(programPath);
      },
    });
    this.code = recast.print(this.ast).code;
    this.ast = recast.parse(this.code, options);
  }
}
