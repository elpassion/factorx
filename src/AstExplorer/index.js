// @flow

import * as babel from 'babel-core';
import Position from '../Position';
import Expression from '../Expression';
import ExpressionNotFoundError from '../ExpressionNotFoundError';
import options from './options';

export default class AstExplorer {
  code: string;
  ast: Object;
  map: Object;
  stop: Function;

  constructor(input: string) {
    const { code, map, ast } = babel.transform(input, options);
    this.code = code;
    this.map = map;
    this.ast = ast;
  }

  serializeNode = (node: Object) => Expression.fromNode(this.code, node);

  findExpressions(selection: Position): Array<Expression> {
    const paths = [];
    this.transform(() => ({
      visitor: {
        Expression(path) {
          if (selection.includes(path.node)) {
            paths.push(path);
          }
        },
      },
    }));
    if (paths.length === 0) throw new ExpressionNotFoundError();
    let parentBlock = paths[paths.length - 1].scope.block;
    if (parentBlock.type === 'ArrowFunctionExpression') parentBlock = parentBlock.body;
    return paths
      .filter(path => parentBlock.start <= path.node.start && parentBlock.end >= path.node.end)
      .map(path => path.node)
      .map(this.serializeNode);
  }

  extractVariable(selection: Position): string {
    let extracted = false;
    this.transform(({ types }) => ({
      visitor: {
        Expression: (path) => {
          const { node } = path;
          if (!node.visited && selection.includes(node)) {
            node.visited = true;
            const id = path.scope.generateUidIdentifierBasedOnNode(node.id);
            path.scope.push({ id, init: path.node });
            path.replaceWith(types.identifier(id.name));
            extracted = true;
          }
        },
      },
    }));
    if (!extracted) throw new ExpressionNotFoundError();
    return this.code;
  }

  transform(plugin: Function) {
    const { code, map, ast } = babel.transformFromAst(this.ast, this.code, {
      ...options,
      plugins: [plugin],
    });
    this.code = code;
    this.map = map;
    this.ast = ast;
  }
}
