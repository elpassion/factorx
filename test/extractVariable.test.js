// @flow
import ExpressionNotFoundError from '../src/ExpressionNotFoundError';
import AstExplorer from '../src/AstExplorer';
import Position from '../src/Position';

describe('extractVariable', () => {
  const expectExtractVariable = (code, start, end) => {
    const astExplorer = new AstExplorer(code);
    expect(astExplorer.extractVariable(new Position(start, end))).toMatchSnapshot();
  };
  describe('with correct selection', () => {
    test('returns correct code', () => {
      expectExtractVariable('5 + 2', 0, 1);
      expectExtractVariable('5 + 2', 4, 5);
      expectExtractVariable('5 + 2', 0, 5);
      expectExtractVariable('() => {\n  5 + 2\n}', 10, 11);
    });
  });

  describe('when incorrect selection', () => {
    test('throws an error', () => {
      expect(() => {
        new AstExplorer('5 + 2').extractVariable(new Position(42, 42));
      }).toThrowError(new ExpressionNotFoundError());
    });
  });

  describe('with multiple selections', () => {
    describe('with correct selection', () => {
      const expectExtractMultipleVariables = (code, selections) => {
        const positions = selections.map(selection => new Position(selection.start, selection.end));
        const astExplorer = new AstExplorer(code);
        expect(astExplorer.extractMultipleVariables(positions)).toMatchSnapshot();
      };
      test('returns correct code', () => {
        expectExtractMultipleVariables('5 + 5', [{ start: 0, end: 1 }, { start: 4, end: 5 }]);
      });
    });

    describe('when incorrect selection', () => {
      test('throws an error', () => {
        const positions = [{ start: 0, end: 1 }, { start: 42, end: 42 }].map(
          selection => new Position(selection.start, selection.end),
        );
        const astExplorer = new AstExplorer('5 + 5');
        expect(() => {
          astExplorer.extractMultipleVariables(positions);
        }).toThrowError(new ExpressionNotFoundError());
      });
    });
  });
});
