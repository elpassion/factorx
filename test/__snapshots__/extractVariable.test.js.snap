// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extractVariable with correct selection returns correct code 1`] = `
Object {
  "code": "var _ref = 5;
_ref + 2",
  "identifierPosition": Position {
    "end": 8,
    "start": 4,
  },
}
`;

exports[`extractVariable with correct selection returns correct code 2`] = `
Object {
  "code": "var _ref = 2;
5 + _ref",
  "identifierPosition": Position {
    "end": 8,
    "start": 4,
  },
}
`;

exports[`extractVariable with correct selection returns correct code 3`] = `
Object {
  "code": "var _ref = 5 + 2;
_ref",
  "identifierPosition": Position {
    "end": 8,
    "start": 4,
  },
}
`;

exports[`extractVariable with correct selection returns correct code 4`] = `
Object {
  "code": "() => {
  var _ref = 5;
  _ref + 2
}",
  "identifierPosition": Position {
    "end": 18,
    "start": 14,
  },
}
`;

exports[`extractVariable with multiple selections with correct selection returns correct code 1`] = `
Object {
  "code": "var _ref = 5;
_ref + _ref",
  "identifierPosition": Position {
    "end": 8,
    "start": 4,
  },
}
`;
